Create VDS if h5file scheme is used

mrosanes opened the pull request:
Create Virtual Datasets if h5file:// scheme is used.
VDS are used to store data coming from 2D detectors, into h5 files.

Commit:
Create VDS if h5file scheme is used

Create Virtual Datasets if h5file:// scheme is used.
VDS are used to store data coming from 2D detectors, into h5 files.

reszelaz reviewed (changes requested):
1. Try to avoid (if possible) opening the source file.
2. If not possible to avoid opening the file, try to avoid loading the whole dataset in memory.
  * get shape from column desc
  * get dtype from column desc (will require major changes - avoid use of str in dtype)
3. Try to handle case when creation of VDS was not possible (try-except)

Minor suggestions:
* coherent variable names (col and column)
* record usually refers to the whole row not a value of a single column
* avoid 4 levels of nesting if/for

Commit:
Minor fixes

- Coherent variable names (col and column)
- Use 'dataset' variable name instead of 'records'. 'Records' usually
refers to the whole row not a value of a single column/channel
- Avoid 4 levels of nesting if/for

Commit:
Do not load whole dataset in memory

Modify code in order to avoid loading the whole h5file://
references dataset in memory (when using python3).

'range' is used instead of 'xrange' thinking that we are going
into the python3 direction.

Commit:
Handle case if it is not possible to create VDS

Use try/except/else to handle case in which it is not possible
to create a VDS. In this case, a backup dataset is kept, with
the h5file:// string references.

reszelaz committed:
Skip creation of VDS and test if h5py does not allow VDS creation

reszelaz committed:
Compare strings with == instead of is

reszelaz committed:
Rename internal variables to be coherent with the rest of the class

reszelaz commented on src/sardana/macroserver/recorders/h5storage.py:
```diff
@@ -325,6 +330,54 @@ def _endRecordList(self, recordlist):
+                bk_name = "_" + name
+                measurement[bk_name] = measurement[name]
+                nb_points = measurement[name].size
+                filename = first_reference.split("://")[1]
```
I don't know if `://` is a correct split. One may ommit `//`.

reszelaz commented on src/sardana/macroserver/recorders/h5storage.py:
```diff
@@ -325,6 +330,54 @@ def _endRecordList(self, recordlist):
+                for i in range(nb_points):
+                    reference = measurement[name][i]
+                    filename = reference.split("://")[1]
+                    vsource = h5py.VirtualSource(filename, "dataset",
```
"dataset" name of the _dataset_ is just an example from the test. We need to extract the real dataset name from the URI, based on `::` split. Regular expressions may be useful.

reszelaz reviewed (changes requested):
I have added check if VDS is available in h5py.
I have modified string comparison and some variable names (minor).
There are still changes to be done.

Commit:
Update test with dataset in URI

reszelaz commented:
To extract the shape and the data type:
```
> /home/zreszela/workspace/sardana/src/sardana/macroserver/recorders/h5storage.py(159)_startRecordList()
-> dd = dd.clone()
(Pdb) dd.name
'tango://pc255.cells.es:10000/expchan/twodctrl03/1'
(Pdb) dd.shape
[4, 4]
(Pdb) dd.dtype
'float64'
```
Is is enough?

Commit:
Use regular expressions instead of split

Use regular expressions to extract scheme, filepath and
dataset.

Commit:
merge sep2 branch in PR

Commit:
Unify all datadescs to have value_ref_enabled flag

Commit:
Use value_ref_enabled instead of dtype

Use value_ref_enabled instead of dtype to determine if
referencing is in use.

dtype contains the data type of the values of the
spectrums or images.

Commit:
Remove expectedFailure decorator for VDS test

Commit:
Fix permits

Commit:
Fix variable using dd.value_ref_enabled

Use data description value_ref_enabled

Commit:
Fix shape on test

Commit:
Use shape from data description

Use shape from channel data description, instead of
extracting the shape from one of the source files.

Commit:
Merge branch 'sep2' into vds

Commit:
Verify if value_ref_enabled is defined

Commit:
Read dtype from data description

Commit:
Fix flake8

Commit:
Fix PEP8 using flake8

Commit:
Improve regular expression for scheme

Commit:
Add TODO for using dataset values

Add TODO for using directly the dataset values, instead of the
reference (filename, etc), in order to identify potential
shape mismatch issues.

Commit:
Use dataset values instead of filename and ds name

Use dataset values as virtual source, instead of filename,
dataset name and shape, in order to avoid possible problems
between source shape and virtual source shape.

Commit:
Revert "Use dataset values instead of filename and ds name"

This reverts commit 33712581816946268501a4ebc2e2cb89d6eac842.

Decided to use filename reference instead of dataset value; to
avoid possible problems if source files are still not present.

Commit:
Revert "Add TODO for using dataset values"

This reverts commit 31cefc94f941fb4d71eef63daefa69574857725a.

Remove TODO. Source file reference will be used instead of
actual dataset, to avoid possible problems, if source file
is still not created, when accessing to the VDS.
